### PR TITLE
ADD: data model and store logic

### PR DESCRIPTION
**Pull request by Argeric**
Define the data model and data storage logic

- - -
This change is [Reviewable](https://reviewable.io/reviews/Conflux-Chain/neurahive-scan/1)

### CODE REVIEW

File: store/store_hex40.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+func (hs *hex40Store) Add(dbTx *gorm.DB, data string, blockTime *time.Time) (uint64, error) {
+	hex := strings.ToLower(strings.TrimPrefix(data, "0x"))
+
+	var hex40 Hex40
+	existed, err := hs.Exists(&hex40, "hex = ?", hex)
+	if err != nil {
+		return 0, err
+	}
```
Comment (wanliqun) [inline store/store_hex40.go:36]: Using LRU cache could improve the query performance. You can add some `TODO` or `comment` for that.

File: store/store_tx.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+			Status:      tx.Status,
+			CreatedAt:   tx.CreatedAt,
+		})
+	}
+
+	return addrTx
+}
+
+func (AddressTx) TableName() string {
+	return "address_tx"
```
Comment (wanliqun) [inline store/store_tx.go:89]: ditto

File: store/types.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package store
+
+import "github.com/pkg/errors"
+
+var (
+	ErrNotFound     = errors.New("not found")
+	ErrChainReorged = errors.New("chain re-orged")
+)
```
Comment (wanliqun) [inline store/types.go:8]: Since we don't have too many types here, maybe we can just move them to `store.go`.

**Review by wanliqun — CHANGES_REQUESTED** [review]
NB Project is not initialized as `go.mod` is missing.

File: store/store.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		if err := ms.submitStore.Add(dbTx, data); err != nil {
+			return errors.WithMessage(err, "failed to save submit event logs")
+		}
+
+		return nil
+	})
+}
+
+func (ms *MysqlStore) Pop(block uint64) error {
+	return nil
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store.go:63]: added

[the remark this answers is not in the document]
Comment (Argeric) [reply store/store.go:63]: > maybe you should add some `TODO` here

added

File: store/store_block.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	blockTime := time.Unix(int64(data.Timestamp), 0)
+	return &Block{
+		BlockNumber: data.Number.Uint64(),
+		Hash:        data.Hash.String()[2:],
+		CreatedAt:   &blockTime,
+	}
+}
+
+func (Block) TableName() string {
+	return "block"
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_block.go:26]: > `Gorm` has a convention to use plural form as table name. Or you can leave out this method.

fixed

File: store/store_base.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	if err == nil {
+		return true, nil
+	}
+
+	if bs.IsRecordNotFound(err) {
+		return false, nil
+	}
+
+	return false, err
+}
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_base.go:40]: > This `BaseStore` seems not to be necessary at all. Try to reuse `https://github.com/Conflux-Chain/go-conflux-util/blob/main/store/mysql/store.go` instead.

fixed

File: store/store.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store.go:63]: fixed

File: store/store_base.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_base.go:40]: fixed

File: store/store_block.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_block.go:26]: fixed

File: store/store_hex40.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
+package store
+
+import (
+	"gorm.io/gorm"
+	"strings"
+	"time"
+)
+
+type Hex40 struct {
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_hex40.go:9]: fixed

File: store/store_hex40.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	Id        uint64     `gorm:"primary_key"`
+	Hex       string     `gorm:"type:varchar(40);unique:idx_hex"`
+	CreatedAt *time.Time `gorm:"not null;index:idx_createdAt,sort:desc"`
+}
+
+func (Hex40) TableName() string {
+	return "hex40"
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_hex40.go:16]: fixed

File: store/store_tx.go (hunk 2 above)
Comment (Argeric) [reply store/store_tx.go:89]: fixed

File: store/types.go (hunk 3 above)
Comment (Argeric) [reply store/types.go:8]: fixed

File: store/store_tx.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+		DripValue:   tx.Value.Uint64(),
+		GasPrice:    tx.GasPrice.Uint64(),
+		Gas:         tx.Gas,
+		Status:      *receipt.Status,
+		CreatedAt:   blockTime,
+	}
+}
+
+func (Tx) TableName() string {
+	return "tx"
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_tx.go:43]: fixed

File: store/store_hex40.go (hunk 1 above)
Comment (Argeric) [reply store/store_hex40.go:36]: fixed

File: store/store_hex40.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	var hex40 Hex40
+	existed, err := hs.Exists(&hex40, "hex = ?", hex)
+	if err != nil {
+		return 0, err
+	}
+
+	if !existed {
+		hex40 = Hex40{
+			Hex:       hex,
+			CreatedAt: blockTime,
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_hex40.go:41]: no，we need accurate blocktime to do some statistic work

File: store/store_log_submit.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+		StartPos:         flowSubmit.StartPos.Uint64(),
+		Length:           flowSubmit.Length.Uint64(),
+		SubmissionLength: flowSubmit.Submission.Length.Uint64(),
+	}
+
+	return submit, nil
+}
+
+func (Submit) TableName() string {
+	return "submit"
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:53]: fixed

File: store/store_log_submit.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+		Identity:         submit.Identity,
+		SubmissionIndex:  submit.SubmissionIndex,
+		StartPos:         submit.StartPos,
+		Length:           submit.Length,
+		SubmissionLength: submit.SubmissionLength,
+	}
+}
+
+func (AddressSubmit) TableName() string {
+	return "address_submit"
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:87]: fixed

File: store/store_tx.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+	From        string     `gorm:"-"`
+	FromId      uint64     `gorm:"not null"`
+	Nonce       uint64     `gorm:"not null"`
+	To          string     `gorm:"-"`
+	ToId        uint64     `gorm:"not null"`
+	DripValue   uint64     `gorm:"not null;default:0"`
+	GasPrice    uint64     `gorm:"not null;default:0"`
+	Gas         uint64     `gorm:"not null;default:0"`
+	Status      uint64     `gorm:"not null"`
+	InterfaceId string     `gorm:"type:varchar(10);not null"`
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_tx.go:22]: fixed

File: store/store_hex40.go [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+		if dbTx != nil {
+			if err := dbTx.Create(&hex40).Error; err != nil {
+				return 0, err
+			}
+		} else {
+			if err := hs.db.Create(&hex40).Error; err != nil {
+				return 0, err
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_hex40.go:52]: fixed

File: store/store_hex40.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+		}
+		if dbTx != nil {
+			if err := dbTx.Create(&hex40).Error; err != nil {
+				return 0, err
+			}
+		} else {
+			if err := hs.db.Create(&hex40).Error; err != nil {
+				return 0, err
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_hex40.go:51]: yes, fixed

File: store/store_log_submit.go [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+
+func newSubmit(blockTime *time.Time, receipt *types.Receipt, log *types.Log, txIndex, txLogIndex int) (*Submit, error) {
+	contract, _ := contract.NewFlowFilterer(common.HexToAddress(""), nil)
+	flowSubmit, err := contract.ParseSubmit(*log.ToEthLog())
+	if err != nil {
+		return nil, err
+	}
+
+	submit := &Submit{
+		BlockNumber:      receipt.BlockNumber,
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:36]: yes, fixed

File: store/store_log_submit.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+		TxPosition:       uint16(txIndex),
+		TxLogPosition:    uint16(txLogIndex),
+		Contract:         log.Address.String()[2:],
+		CreatedAt:        blockTime,
+		Sender:           log.Topics[1].String()[26:],
+		Identity:         log.Topics[2].String()[2:],
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:42]: fixed

File: store/store_log_submit.go [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+		if receipt == nil || !IsTxExecutedInBlock(&tx, receipt) {
+			continue
+		}
+		for k, log := range receipt.Logs {
+			contract := log.Address.String()
+			topic0 := log.Topics[0].String()
+			if !strings.EqualFold(contract, ss.sc.FlowContract) || topic0 != ss.sc.SubmitEventSignature {
+				continue
+			}
+
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:120]: fixed

File: store/store_log_submit.go [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+	}
+}
+
+func (AddressSubmit) TableName() string {
+	return "address_submit"
+}
+
+type submitStore struct {
+	hs *hex40Store
+	sc *SyncConfig
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:92]: fixed

File: store/store_log_submit.go [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+import (
+	"github.com/Conflux-Chain/neurahive-client/contract"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/openweb3/web3go/types"
+	"gorm.io/gorm"
+	"strings"
+	"time"
+)
+
+type Submit struct {
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_log_submit.go:12]: fixed

File: store/store.go [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+type SyncConfig struct {
+	FlowContract         string
+	SubmitEventSignature string
+}
+
+func MustNewStore(db *gorm.DB) *MysqlStore {
+	var sc SyncConfig
+	viper.MustUnmarshalKey("sync", &sc)
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store.go:29]: fixed

File: store/store.go [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+	*txStore
+	*submitStore
+	*hex40Store
+	syncConfig *SyncConfig
+}
+
+type SyncConfig struct {
+	FlowContract         string
+	SubmitEventSignature string
+}
```
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store.go:25]: fixed

File: store/store_hex40.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Argeric) [reply store/store_hex40.go]: already added blockTime and createdAt

**Review by wanliqun — CHANGES_REQUESTED** [review]
> Reviewed 1 of 8 files at r1, 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1)* status: all files reviewed, 1 unresolved discussion (waiting on @Argeric and @Tigo9527)

___
*[`store/store.go` line 25 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311045596:-Nd9dbZOCbtNNnGDR14g:b-hvb0vg) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store.go#L25)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

I think you can also delete this struct as well since `SyncConfig` shouldn't be here.
`type SyncConfig struct {
	FlowContract         string
	SubmitEventSignature string
}`
___
*[`store/store.go` line 29 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311047244:-Nd9ghvA7MGFRQIPkW_a:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store.go#L29)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store.go` line 63 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311045815:-Nd9e66D-lj0w8-d7-Pl:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store.go#L63)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_block.go` line 26 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311050942:-Nd9eJ4P1wA4RYv56bmI:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_block.go#L26)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_tx.go` line 22 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311125307:-Nd9hr3h6dAk2tKfpPfa:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_tx.go#L22)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_tx.go` line 43 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311125801:-Nd9ejzx9wrPgeLbCnzB:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_tx.go#L43)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_tx.go` line 89 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311134473:-Nd9elAt-Me9F1Yis-TT:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_tx.go#L89)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_base.go` line 40 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311046335:-Nd9fLkB1QUi3OdLIVEh:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_base.go#L40)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_hex40.go` line 9 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311053323:-Nd9er3d2uNH7K0zP9c1:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_hex40.go#L9)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_hex40.go` line 16 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311053537:-Nd9esFx2zafnVjJn9ec:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_hex40.go#L16)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_hex40.go` line 36 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311131927:-Nd9f_Na9WZ0MagsE2sc:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_hex40.go#L36)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_hex40.go` line 51 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311057515:-Nd9fOuS4fcX6XZJuAma:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_hex40.go#L51)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

yes, fixed
</blockquote></details>

LGTM
___
*[`store/store_hex40.go` line 52 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311055204:-Nd9fNpN6ucrQhpKod2l:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_hex40.go#L52)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/types.go` line 8 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311142068:-Nd9feZ6A8LVUhtUF7pM:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/types.go#L8)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 12 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311116245:-Nd9bBx0ApQZFLUNziwn:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L12)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 36 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311058390:-Nd9ePTv-IGw-Hgzk6Ap:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L36)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

yes, fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 42 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311061513:-Nd9eQp4Ed3J4zYvOFRD:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L42)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 53 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311116782:-Nd9eSBK9WvS3_2hRYEO:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L53)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 87 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311120923:-Nd9eUEi4JvXlDdtQeOZ:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L87)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 92 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311121553:-Nd9eX2l3akiCorr_l1F:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L92)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM
___
*[`store/store_log_submit.go` line 120 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311123376:-Nd9eZwM3uUDc_ZJ_eZj:bb6ywmx) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store_log_submit.go#L120)):*
<details><summary><i>Previously, Argeric wrote…</i></summary><blockquote>

fixed
</blockquote></details>

LGTM

**Review by Argeric — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1)* status: 9 of 10 files reviewed, 1 unresolved discussion (waiting on @Tigo9527 and @wanliqun)

___
*[`store/store.go` line 25 at r1](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#gh-1311045596:-Nd9j_Z57NfarGQmodAj:b-7ql1td) ([raw file](https://github.com/conflux-chain/neurahive-scan/blob/30d206448cc12a8a3801209d45586678ff597a90/store/store.go#L25)):*
<details><summary><i>Previously, wanliqun (wanliqun) wrote…</i></summary><blockquote>

I think you can also delete this struct as well since `SyncConfig` shouldn't be here.
`type SyncConfig struct {
	FlowContract         string
	SubmitEventSignature string
}`
</blockquote></details>

fixed

**Review by wanliqun — APPROVED** [review]
LGTM

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/neurahive-scan/1#-:-Nd9jnez2f8zKBz0EGZg:bb6ywmx)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Tigo9527)

**Pull request merged**